basic_test.go

```diff
@@ -363,69 +363,89 @@ func TestBothMany(t *testing.T) {
 	testMany(t, 10000, 100, 10, false)
 }
 
-func BenchmarkSubs(b *testing.B) {
-	b.ReportAllocs()
-	testMany(b, b.N, 100, 100, false)
+type benchCase struct {
+	subs     int
+	emits    int
+	stateful bool
 }
 
-func BenchmarkEmits(b *testing.B) {
-	b.ReportAllocs()
-	testMany(b, 100, b.N, 100, false)
+func (bc benchCase) name() string {
+	return fmt.Sprintf("subs-%d/emits-%d/stateful-%t", bc.subs, bc.emits, bc.stateful)
 }
 
-func BenchmarkMsgs(b *testing.B) {
-	b.ReportAllocs()
-	testMany(b, 100, 100, b.N, false)
-}
-
-func BenchmarkOneToMany(b *testing.B) {
-	b.ReportAllocs()
-	testMany(b, b.N, 1, 100, false)
-}
-
-func BenchmarkManyToOne(b *testing.B) {
-	b.ReportAllocs()
-	testMany(b, 1, b.N, 100, false)
-}
-
-func BenchmarkMs1e2m4(b *testing.B) {
-	b.N = 1000000
-	b.ReportAllocs()
-	testMany(b, 10, 100, 10000, false)
+func genTestCases() []benchCase {
+	ret := make([]benchCase, 0, 200)
+	for stateful := 0; stateful < 2; stateful++ {
+		for subs := uint(0); subs <= 8; subs = subs + 2 {
+			for emits := uint(0); emits <= 8; emits = emits + 2 {
```

Comment:
I'd do 
```suggestion
			for emits := uint(0); emits <= 8; emits = emits + 4 {
```
On both of these.